README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
fiftystater
===========

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/fiftystater)](https://cran.r-project.org/package=fiftystater) [![Travis-CI Build Status](https://travis-ci.org/wmurphyrd/fiftystater.svg?branch=master)](https://travis-ci.org/wmurphyrd/fiftystater)

This data package provides an easy way to plot 50-state choropleth thematic maps with ggplot2 in R.

Motivation
----------

It is far too difficult to create a U.S. state data visualization in R that includes Alaska and Hawaii. The U.S. state map included in the `maps` package includes only the 48 contiguous states. A search for existing solutions will lead to Stack Overflow threads ([1](http://stackoverflow.com/questions/13757771/), [2](http://stackoverflow.com/questions/25530358/)), but the solutions available require either limiting plotting options by rendering with a specific package, installing a large list of dependencies and tracking down shape files to perform arcane transformation calculations, or using annotations or visibility-based maps that compromise the aesthetic of the visualization.

This package is a simple solution: a data file, with no dependencies, that can be dropped into ggplot2 with no effort. To create it, I followed the steps in a [Stack Overflow answer](http://stackoverflow.com/a/13767984) from Barry Rowlingson to transform Alaska and Hawaii into insets next to the contiguous states and packaged the resulting data frame for simple drop-in use.

Installation
------------

You can install fiftystater from github with:

``` r
# install.packages("devtools")
devtools::install_github("wmurphyrd/fiftystater")
```

#### Version 1.0.1

-   `fifty_states` data object to easily make 50-state choropleth thematic maps in ggplot2 by including Alaska and Hawaii as insets
-   `fifty_states_inset_boxes` function to add border boxes to the insets, if desired

Usage Examples
--------------

To use the fifty state map, simply point the `map` argument in `geom_map` to `fifty_states`. As usual, you'll also create an aesthetic mapping for `map_id` to the column in your data frame that contains the lower case state names.

``` r
library(ggplot2)
library(fiftystater)

data("fifty_states") # this line is optional due to lazy data loading

crimes <- data.frame(state = tolower(rownames(USArrests)), USArrests)

# map_id creates the aesthetic mapping to the state name column in your data
p <- ggplot(crimes, aes(map_id = state)) + 
  # map points to the fifty_states shape data
  geom_map(aes(fill = Assault), map = fifty_states) + 
  expand_limits(x = fifty_states$long, y = fifty_states$lat) +
  coord_map() +
  scale_x_continuous(breaks = NULL) + 
  scale_y_continuous(breaks = NULL) +
  labs(x = "", y = "") +
  theme(legend.position = "bottom", 
        panel.background = element_blank())

p
```

![](README-example-1.png)

If you need to be highly technically accurate, you can add bounding boxes to highlight the insets so that no viewer confuses Alaska and Hawaii for regions in a deformed Mexico that you randomly decided to include. The function called `fifty_states_inset_boxes` will return a layer created via `annotate`, so you can just add it to your plot.

``` r
# add border boxes to AK/HI
p + fifty_states_inset_boxes()
```

![](README-example_box-1.png)

This package also plays nicely with my [colorplaner](https://cran.r-project.org/package=colorplaner) package for bivariate color scales.

``` r
# Map a second variable to each state's fill color with colorplaner
library(colorplaner)
p + aes(fill2 = UrbanPop) + scale_fill_colorplane() +
  theme(legend.position = "right")
```

![](README-example_colorplaner-1.png)

The map data included in this package can be generated with the code below. If you want to run this code, you will need to ensure the packages listed below are installed; they are not automatically installed with `fiftystater`.

``` r
# Create map data with AK, HI inset.
# Inspired by http://stackoverflow.com/a/13767984
library(rgeos)
library(maptools)
library(rgdal)
library(ggplot2)
library(dplyr)

transform_state <- function(object, rot, scale, shift){
  object %>% elide(rotate = rot) %>%
    elide(scale = max(apply(bbox(object), 1, diff)) / scale) %>%
    elide(shift = shift)
}

#state shape file from
# http://www.arcgis.com/home/item.html?id=f7f805eb65eb4ab787a0a3e1116ca7e5
loc <- file.path(tempdir(), "stats_dat")
unzip("inst/extdata/states_21basic.zip", exdir = loc)
fifty_states_sp <- readOGR(dsn = loc, layer = "states", verbose = FALSE) %>%
  spTransform(CRS("+init=epsg:2163"))

alaska <- fifty_states_sp[fifty_states_sp$STATE_NAME == "Alaska", ] %>%
  transform_state(-35, 2.5, c(-2400000, -2100000))
proj4string(alaska) <- proj4string(fifty_states_sp)

hawaii <- fifty_states_sp[fifty_states_sp$STATE_NAME == "Hawaii", ] %>%
  transform_state(-35, .75, c(-1170000,-2363000))
proj4string(hawaii) <- proj4string(fifty_states_sp)

fifty_states <-
  fifty_states_sp[!fifty_states_sp$STATE_NAME %in% c("Alaska","Hawaii"), ] %>%
  rbind(alaska) %>%
  rbind(hawaii) %>%
  spTransform(CRS("+init=epsg:4326")) %>%
  fortify(region = "STATE_NAME") %>%
  mutate(id = tolower(id))

devtools::use_data(fifty_states, overwrite = TRUE)

get_box <- function(id) {
  fifty_states[fifty_states$id == id, c("long", "lat")] %>%
    as.matrix %>%
    bbox %>%
    t %>%
    # expand slightly to leave inner margin
    apply(2, `+`, c(-.5, .5)) %>%
    # rearrange corner coordinates into path coordinates
    `[`(c(1, 2, 2, 1, 1, 3, 3, 4, 4, 3)) %>%
    matrix(ncol = 2) %>%
    as.data.frame %>%
    setNames(c("x", "y")) %>%
    cbind(data.frame(id = id, stringsAsFactors = FALSE))
}

fifty_states_inset_boxes_data <- c("alaska", "hawaii") %>%
  lapply(get_box) %>%
  bind_rows

devtools::use_data(fifty_states_inset_boxes_data, overwrite = TRUE)
```


NEWS.md
#### Version 1.0.1

* `fifty_states` data object to easily make 50-state choropleth thematic maps in ggplot2 by including Alaska and Hawaii as insets
* `fifty_states_inset_boxes` function to add border boxes to the insets, 
if desired


tests/testthat.R
library(testthat)
library(fiftystater)

test_check("fiftystater")


tests/testthat/test_fifty_states.R
library(ggplot2)

test_that("Data can be loaded and used", {
  crimes <- data.frame(state = tolower(rownames(USArrests)), USArrests)
  expect_silent({
    print(
      ggplot(crimes, aes(map_id = state)) +
        geom_map(aes(fill = Murder), map = fifty_states) +
        expand_limits(x = fifty_states$long, y = fifty_states$lat) +
        coord_map()
    )
  })
})

test_that("Inset borders can be loaded and used", {
  crimes <- data.frame(state = tolower(rownames(USArrests)), USArrests)
  expect_silent({
    print(
      ggplot(crimes, aes(map_id = state)) +
        geom_map(aes(fill = Murder), map = fifty_states) +
        expand_limits(x = fifty_states$long, y = fifty_states$lat) +
        coord_map() + fifty_states_inset_boxes()
    )
  })
})


R/fiftystater.R
# This file is part of fiftystater.
#
# fiftystater is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3.
#
# fiftystater is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with fiftystater  If not, see <http://www.gnu.org/licenses/>.

#' Fifty U.S. States and Washington D.C. with Alaska and Hawaii as Insets
#'
#' Map data to plot the entire United States by transforming Alaska and Hawaii
#' to appear as insets nearby the continental states.
#'
#' @format A data frame with 13694 rows and 7 variables:
#' \describe{
#'   \item{long,lat}{Numeric, latitude and longitude coordinates for shapes.}
#'   \item{order}{Integer, drawing order of shapes.}
#'   \item{hole}{Logical, whether the shape represents a negation.}
#'   \item{piece}{Factor, groupings of shapes.}
#'   \item{id}{Character, lower case state names mapped to plot data via
#'      the \code{map_id} aesthetic in \code{\link[ggplot2]{geom_map}}.}
#'   \item{group}{Factor, groupings of shapes.}
#' }
#'
#' @source Original shapefile provided by Charlie Fitzpatrick
#' \href{http://www.arcgis.com/home/item.html?id=f7f805eb65eb4ab787a0a3e1116ca7e5}{via ArcGIS}.
#' Transformations based on code from Barry Rowlingson on StackOverflow
#' \url{http://stackoverflow.com/a/13767984}.
#'
#' @details The \code{fifty_states} data frame is provided in the format
#'   expected by the \code{map} argument to \code{\link[ggplot2]{geom_map}}.
#'   Border boxes for the inset states can be included by adding
#'   \code{\link{fifty_states_inset_boxes}} to the plot.
#'
#' @examples
#' if(require("mapproj") && require("ggplot2")) {
#'   crimes <- data.frame(state = tolower(rownames(USArrests)), USArrests)
#'   p <- ggplot(crimes, aes(fill = UrbanPop, map_id = state)) +
#'     geom_map(map = fifty_states) +
#'     expand_limits(x = fifty_states$long, y = fifty_states$lat) +
#'     coord_map()
#'  p
#'  # Add bounding boxes for Alaska and Hawaii
#'  p + fifty_states_inset_boxes()
#' }
#'
#' @seealso \code{\link{fifty_states_inset_boxes}}
"fifty_states"

#' Add inset box borders
#'
#' Border boxes for the insets of Alaska and Hawaii in
#' \code{\link{fifty_states}} can be added to the map with this function.
#'
#' Adding the result of this function to your ggplot will add a layer with
#' border boxes to highlight the insets. It draws data from the package data
#' object
#' \code{fifty_states_inset_boxes_data} and creates a layer to plot the borders
#'  with \code{\link[ggplot2]{annotate}}.
#'
#'
#' @return A ggplot2 layer object with path data to draw the inset border boxes
#'
#' @examples
#' if(require("mapproj") && require("ggplot2")) {
#'   crimes <- data.frame(state = tolower(rownames(USArrests)), USArrests)
#'   ggplot(crimes, aes(fill = UrbanPop, map_id = state)) +
#'     geom_map(map = fifty_states) +
#'     expand_limits(x = fifty_states$long, y = fifty_states$lat) +
#'     coord_map() + fifty_states_inset_boxes()
#' }
#' @aliases fifty_states_inset_boxes_data
#' @export
fifty_states_inset_boxes <- function() {
  if(!requireNamespace("ggplot2")) {
    stop("Package 'ggplot2' required for inset box annotation.")
  }
  ggplot2::annotate("path",
                    x = fiftystater::fifty_states_inset_boxes_data$x,
                    y = fiftystater::fifty_states_inset_boxes_data$y,
                    group = fiftystater::fifty_states_inset_boxes_data$id)
}



data-raw/make_mapfile.R
# This file is part of fiftystater.
#
# fiftytater is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3.
#
# fiftystater is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with fiftystater  If not, see <http://www.gnu.org/licenses/>.

# Create map data with AK, HI inset.
# Inspired by http://stackoverflow.com/a/13767984
library(rgeos)
library(maptools)
library(rgdal)
library(ggplot2)
library(dplyr)

transform_state <- function(object, rot, scale, shift){
  object %>% elide(rotate = rot) %>%
    elide(scale = max(apply(bbox(object), 1, diff)) / scale) %>%
    elide(shift = shift)
}

#state shape file from
# http://www.arcgis.com/home/item.html?id=f7f805eb65eb4ab787a0a3e1116ca7e5
loc <- file.path(tempdir(), "stats_dat")
unzip("inst/extdata/states_21basic.zip", exdir = loc)
fifty_states_sp <- readOGR(dsn = loc, layer = "states", verbose = FALSE) %>%
  spTransform(CRS("+init=epsg:2163"))

alaska <- fifty_states_sp[fifty_states_sp$STATE_NAME == "Alaska", ] %>%
  transform_state(-35, 2.5, c(-2400000, -2100000))
proj4string(alaska) <- proj4string(fifty_states_sp)

hawaii <- fifty_states_sp[fifty_states_sp$STATE_NAME == "Hawaii", ] %>%
  transform_state(-35, .75, c(-1170000,-2363000))
proj4string(hawaii) <- proj4string(fifty_states_sp)

fifty_states <-
  fifty_states_sp[!fifty_states_sp$STATE_NAME %in% c("Alaska","Hawaii"), ] %>%
  rbind(alaska) %>%
  rbind(hawaii) %>%
  spTransform(CRS("+init=epsg:4326")) %>%
  fortify(region = "STATE_NAME") %>%
  mutate(id = tolower(id))

devtools::use_data(fifty_states, overwrite = TRUE)

get_box <- function(id) {
  fifty_states[fifty_states$id == id, c("long", "lat")] %>%
    as.matrix %>%
    bbox %>%
    t %>%
    # expand slightly to leave inner margin
    apply(2, `+`, c(-.5, .5)) %>%
    # rearrange corner coordinates into path coordinates
    `[`(c(1, 2, 2, 1, 1, 3, 3, 4, 4, 3)) %>%
    matrix(ncol = 2) %>%
    as.data.frame %>%
    setNames(c("x", "y")) %>%
    cbind(data.frame(id = id, stringsAsFactors = FALSE))
}

fifty_states_inset_boxes_data <- c("alaska", "hawaii") %>%
  lapply(get_box) %>%
  bind_rows

devtools::use_data(fifty_states_inset_boxes_data, overwrite = TRUE)
